chainer/gradient_check.py

```diff
@@ -235,13 +248,16 @@ def f():
         if len(no_grads) != len(xs):
             raise ValueError(
                 'Length of no_grads param and xs should be same.')
-    for skip, x in six.moves.zip(no_grads, xs):
+    for skip, x, cx in six.moves.zip(no_grads, xs, casted_xs):
         if skip:
             assert x.grad is None
             continue
-        gx, = numerical_grad(f, (x.data,), y_grad, eps=eps)
+        gx, = numerical_grad(f, (cx.data,), y_grad, eps=eps)
         testing.assert_allclose(gx, x.grad, atol=atol, rtol=rtol)
-        assert gx.dtype is x.grad.dtype
+        if dtype is None:
+            assert gx.dtype == x.grad.dtype
+        else:
+            assert gx.dtype == dtype
```

Comment:
I found that when that when the `dtype` of `x` is `int32`, `gx` will be `int32`. As `x` is not cast to `dtype`, whose kind is `float` in the current implementation, the assertion should fail. The possible solutions are

* to convert `x` even if `x` is not float.
* to impose `x` to be float.

Could you consider the problem?